### DISCUSSION TITLE
RandomLongIdentityGenerator vs GuidIdentityGenerator

### DISCUSSION THREAD

**Discussion by ColinGski** [elsa-workflows/elsa-core#5443] [Q&A]
Hello !

I've a question regarding the IDs used in the database by ELSA.
I see that those are generated by an interface called **IIdentityGenerator** for which there are multiple implementations.

The default one is **RandomLongIdentityGenerator** which is producing a random long converted to hexadecimal.
I also see there's a **GuidIdentityGenerator** that can be setup as follows:

`elsa.UseWorkflows(workflows => workflows.WithIdentityGenerator(_ => new GuidIdentityGenerator()));`

My question is: Why is **RandomLongIdentityGenerator** the default ?
The risk of collision is higher with a random long number than with a Guid.

Any reason why I should not use **GuidIdentityGenerator** ?

Best regards.

**Comment by glime-ai** [comment]
The choice of `RandomLongIdentityGenerator` as the default identity generator in Elsa is likely due to a balance between performance and uniqueness. While it's true that GUIDs offer a lower risk of collision due to their larger size and generation algorithm, they can also be more resource-intensive to generate and store due to their 128-bit size compared to a 64-bit long.

The `RandomLongIdentityGenerator` may have been chosen as a sensible default that provides a good trade-off between the likelihood of collision (which is still very low with a random long) and performance/storage considerations. It's also possible that the developers of Elsa considered the typical use cases and scale at which Elsa workflows would be used and determined that the risk of collision was acceptably low.

However, if your specific use case requires the additional uniqueness guarantees provided by GUIDs, or if you simply prefer to use GUIDs for identity generation, you can certainly configure Elsa to use the `GuidIdentityGenerator`. There's no inherent reason why you should not use it if it better suits your needs.

Before making the switch, you might want to consider the implications on your system's performance and storage, as well as whether your application's scale and use case justify the need for GUIDs over long integers.

If you decide to use `GuidIdentityGenerator`, you can set it up as you've shown in your question, and Elsa should work with GUIDs for identity generation without any issues.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
